spring-kafka/src/main/java/org/springframework/kafka/listener/ErrorHandlingUtils.java

```diff
@@ -126,25 +128,35 @@ public static void retryBatch(Exception thrownException, ConsumerRecords<?, ?> r
 					consumer.poll(Duration.ZERO);
 				}
 				catch (WakeupException we) {
-					seeker.handleBatch(thrownException, records, consumer, container, () -> { });
+					seeker.handleBatch(thrownException, records, consumer, container, NO_OP);
 					throw new KafkaException("Woken up during retry", logLevel, we);
 				}
 				try {
-					ListenerUtils.stoppableSleep(container, nextBackOff);
+					ListenerUtils.stoppableSleep(
+							() -> container.isRunning() &&
+									!container.isPauseRequested() &&
+									records.partitions().stream().noneMatch(container::isPartitionPauseRequested),
+							nextBackOff
+					);
 				}
 				catch (InterruptedException e1) {
 					Thread.currentThread().interrupt();
-					seeker.handleBatch(thrownException, records, consumer, container, () -> { });
+					seeker.handleBatch(thrownException, records, consumer, container, NO_OP);
 					throw new KafkaException("Interrupted during retry", logLevel, e1);
 				}
 				if (!container.isRunning()) {
 					throw new KafkaException("Container stopped during retries");
 				}
+				if (container.isPauseRequested() ||
+						records.partitions().stream().anyMatch(container::isPartitionPauseRequested)) {
```

Comment:
Perhaps traversing the records twice per iteration is a bit much (especially with large `max.poll.records`), given that user pausing individual partitions is a corner case (I am only aware of one user that uses this feature - it was added for non-blocking retries, which is not supported for batch listeners).

Maybe we could add `isAnyPartitionPauseRequested()` to the container and just use that here instead?

Reply:
Oops - ignore this; we are not traversing the records, just the partitions.

Reply:
I wouldn't expect that set of TopicPartitions will be that gigantic. Even though max.poll.records could be 500k, all of them could be from dozen partitions.

>  Maybe we could add `isAnyPartitionPauseRequested()` to the container 

Hmmm, feels wrong to rely on it. Because, it could be that some other partition is paused, which none of records ate from that partition.